File: python/app/services/follows.py
"""follows collection: one row per (followerId, followingId) pair, same
check-then-act uniqueness guard as likes.py (see that module's docstring).
`followingName` is denormalized onto the row at write time — see
plan/build-plan.md "Known limitations" for why (no `users` collection to
resolve a display name from otherwise).
"""

import asyncio

from app.config import get_settings
from app.mudbase_client import create_data_sync, delete_data_sync, list_data_sync
from app.schemas.social import Follow


async def find_follow(follower_id: str, following_id: str, *, access_token: str | None = None) -> Follow | None:
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.follows_collection_id,
        filter_dict={"followerId": follower_id, "followingId": following_id},
        limit=1,
        access_token=access_token,
    )
    docs = result["data"]
    return Follow.model_validate(docs[0]) if docs else None


async def count_followers(user_id: str, *, access_token: str | None = None) -> int:
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.follows_collection_id,
        filter_dict={"followingId": user_id},
        limit=1,
        access_token=access_token,
    )
    pagination = result["pagination"]
    return int(pagination["total"]) if pagination else 0


async def count_following(user_id: str, *, access_token: str | None = None) -> int:
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.follows_collection_id,
        filter_dict={"followerId": user_id},
        limit=1,
        access_token=access_token,
    )
    pagination = result["pagination"]
    return int(pagination["total"]) if pagination else 0


async def find_any_follow_naming(user_id: str, *, access_token: str | None = None) -> Follow | None:
    """Used by profiles.py's best-effort display-name resolution: a row
    recorded by anyone who follows `user_id` carries their denormalized
    `followingName`, which is the last fallback before the literal
    string "Member" (see plan/build-plan.md "Known limitations")."""
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.follows_collection_id,
        filter_dict={"followingId": user_id},
        limit=1,
        access_token=access_token,
    )
    docs = result["data"]
    return Follow.model_validate(docs[0]) if docs else None


async def toggle_follow(
    follower_id: str,
    following_id: str,
    *,
    following_name: str | None,
    access_token: str,
) -> bool:
    """Returns the new following state (True if now following, False if now unfollowed)."""
    settings = get_settings()
    existing = await find_follow(follower_id, following_id, access_token=access_token)
    if existing is not None:
        await asyncio.to_thread(
            delete_data_sync, settings.follows_collection_id, existing.id, access_token=access_token
        )
        return False

    await asyncio.to_thread(
        create_data_sync,
        settings.follows_collection_id,
        {"followerId": follower_id, "followingId": following_id, "followingName": following_name},
        access_token=access_token,
    )
    return True

File: go/internal/store/follows.go
package store

import (
	"context"
	"fmt"

	"github.com/mudbase/mudbase-showcase-social/go/internal/mbase"
	"github.com/mudbase/mudbase-showcase-social/go/internal/models"
)

// FollowService implements every operation the app needs against the `follows` collection.
type FollowService struct {
	client       *mbase.Client
	collectionID string
}

// NewFollowService builds a FollowService bound to the given follows collection ID.
func NewFollowService(client *mbase.Client, followsCollectionID string) *FollowService {
	return &FollowService{client: client, collectionID: followsCollectionID}
}

// findFollow looks up the (followerID, followingID) follow row, ok=false if none exists.
func (s *FollowService) findFollow(ctx context.Context, token, followerID, followingID string) (follow models.Follow, ok bool, err error) {
	result, err := mbase.List[models.Follow](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"followerId": followerID, "followingId": followingID},
		Limit:  1,
	})
	if err != nil {
		return models.Follow{}, false, fmt.Errorf("store: looking up follow %s->%s: %w", followerID, followingID, err)
	}
	if len(result.Data) == 0 {
		return models.Follow{}, false, nil
	}
	return result.Data[0], true, nil
}

// IsFollowing reports whether followerID currently follows followingID.
func (s *FollowService) IsFollowing(ctx context.Context, token, followerID, followingID string) (bool, error) {
	_, ok, err := s.findFollow(ctx, token, followerID, followingID)
	return ok, err
}

// Toggle follows followingID for followerID if not already following, or unfollows if already
// following - the same check-then-act tradeoff as LikeService.Toggle. followingName is
// denormalized onto the row (see models.Follow) so followingID's profile can resolve a display
// name even if they've never posted. Returns the new following state.
func (s *FollowService) Toggle(ctx context.Context, token, followerID, followingID, followingName string) (following bool, err error) {
	existing, ok, err := s.findFollow(ctx, token, followerID, followingID)
	if err != nil {
		return false, err
	}

	if ok {
		if err := mbase.Delete(ctx, s.client, token, s.collectionID, existing.ID); err != nil {
			return false, fmt.Errorf("store: removing follow: %w", err)
		}
		return false, nil
	}

	body := map[string]interface{}{
		"followerId":  followerID,
		"followingId": followingID,
	}
	if followingName != "" {
		body["followingName"] = followingName
	}
	if _, err := mbase.Create[models.Follow](ctx, s.client, token, s.collectionID, body); err != nil {
		return false, fmt.Errorf("store: creating follow: %w", err)
	}
	return true, nil
}

// FollowerCount returns how many accounts follow userID, via pagination.Total on a limit-1 query -
// no need to fetch the rows themselves for a plain count.
func (s *FollowService) FollowerCount(ctx context.Context, token, userID string) (int32, error) {
	result, err := mbase.List[models.Follow](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"followingId": userID},
		Limit:  1,
	})
	if err != nil {
		return 0, fmt.Errorf("store: counting followers for %s: %w", userID, err)
	}
	return result.Total, nil
}

// FollowingCount returns how many accounts userID follows.
func (s *FollowService) FollowingCount(ctx context.Context, token, userID string) (int32, error) {
	result, err := mbase.List[models.Follow](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"followerId": userID},
		Limit:  1,
	})
	if err != nil {
		return 0, fmt.Errorf("store: counting following for %s: %w", userID, err)
	}
	return result.Total, nil
}

// AnyFollowingNameFor returns the followingName recorded by anyone who follows userID, ok=false if
// nobody follows them yet or none of their followers' rows carry a name - used as the second
// fallback in ProfileService.ResolveDisplayName.
func (s *FollowService) AnyFollowingNameFor(ctx context.Context, token, userID string) (name string, ok bool, err error) {
	result, err := mbase.List[models.Follow](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"followingId": userID},
		Limit:  20,
	})
	if err != nil {
		return "", false, fmt.Errorf("store: looking up a following name for %s: %w", userID, err)
	}
	for _, follow := range result.Data {
		if follow.FollowingName != "" {
			return follow.FollowingName, true, nil
		}
	}
	return "", false, nil
}
